src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexRunner.cs

```diff
@@ -207,6 +151,86 @@ public abstract class RegexRunner
             }
         }
 
+        internal void InitializeForScan(Regex regex, ReadOnlySpan<char> text, int textstart, bool quick)
+        {
+            this.quick = quick;
+            // Store remaining arguments into fields now that we're going to start the scan.
+            // These are referenced by the derived runner.
+            runregex = regex;
+            runtextstart = textstart;
+            runtextbeg = 0;
+            runtextend = text.Length;
+            runtextpos = textstart;
+
+            if (runmatch is null)
+            {
+                // Use a hashtabled Match object if the capture numbers are sparse
+                runmatch = runregex!.caps is null ?
+                    new Match(runregex, runregex.capsize, runtext, runtextbeg, runtextend - runtextbeg, runtextstart) :
+                    new MatchSparse(runregex, runregex.caps, runregex.capsize, runtext, runtextbeg, runtextend - runtextbeg, runtextstart);
+            }
+            else
+            {
+                runmatch.Reset(runregex!, runtext, runtextbeg, runtextend, runtextstart);
+            }
+
+            // Note we test runcrawl, because it is the last one to be allocated
+            // If there is an alloc failure in the middle of the three allocations,
+            // we may still return to reuse this instance, and we want to behave
+            // as if the allocations didn't occur.
+            if (runcrawl != null)
+            {
+                runtrackpos = runtrack!.Length;
+                runstackpos = runstack!.Length;
+                runcrawlpos = runcrawl.Length;
+                return;
+            }
+
+            // Everything above runs once per match.
+            // Everything below runs once per runner.
+
+            InitTrackCount();
+
+            int stacksize;
+            int tracksize = stacksize = runtrackcount * 8;
+
+            if (tracksize < 32)
+            {
+                tracksize = 32;
+            }
+            if (stacksize < 16)
+            {
+                stacksize = 16;
+            }
+
+            runtrack = new int[tracksize];
+            runtrackpos = tracksize;
+
+            runstack = new int[stacksize];
+            runstackpos = stacksize;
+
+            runcrawl = new int[32];
+            runcrawlpos = 32;
```

Comment:
We don't need to do it now, but we're going to want to revisit this logic.

Reply:
I assume what you are thinking is only do this initialization if the current runner is Interpreted right?

Reply:
Partially.  We only need to initialize runtrack if the runner is interpreted.  We need runstack for Compiled and the source generator as well, but not for NonBacktracking, and the size is also based off of runtrackcount which is largely irrelevant, so we need to figure out what the right default size should be.  runcrawl is only relevant if there are captures, so it's very likely the default size chosen here is too large.